src/crimson/os/seastore/cached_extent.cc

```diff
@@ -77,21 +80,85 @@ CachedExtent::~CachedExtent()
     parent_index->erase(*this);
   }
 }
+CachedExtent* CachedExtent::get_transactional_view(Transaction &t) {
+  return get_transactional_view(t.get_trans_id());
+}
 
-std::ostream &LogicalCachedExtent::print_detail(std::ostream &out) const
-{
-  out << ", laddr=" << laddr;
-  if (pin) {
-    out << ", pin=" << *pin;
+CachedExtent* CachedExtent::get_transactional_view(transaction_id_t tid) {
+  auto it = mutation_pendings.find(tid, trans_spec_view_t::cmp_t());
+  if (it != mutation_pendings.end()) {
+    return (CachedExtent*)&(*it);
+  } else {
+    return this;
+  }
+}
+
+std::ostream &operator<<(std::ostream &out, const parent_tracker_t &tracker) {
+  return out << "parent_tracker=" << (void*)&tracker
+	     << ", parent=" << (void*)tracker.get_parent().get();
+}
+
+std::ostream &ChildableCachedExtent::print_detail(std::ostream &out) const {
+  if (parent_tracker) {
+    out << *parent_tracker;
   } else {
-    out << ", pin=empty";
+    out << ", parent_tracker=" << (void*)nullptr;
   }
+  _print_detail(out);
+  return out;
+}
+
+std::ostream &LogicalCachedExtent::_print_detail(std::ostream &out) const
+{
+  out << ", laddr=" << laddr;
   return print_detail_l(out);
 }
 
-std::ostream &operator<<(std::ostream &out, const LBAPin &rhs)
+void child_pos_t::link_child(ChildableCachedExtent *c) {
+  get_parent<FixedKVNode<laddr_t>>()->link_child(c, pos);
+}
+
+void CachedExtent::set_invalid(Transaction &t) {
+  state = extent_state_t::INVALID;
+  if (trans_view_hook.is_linked()) {
+    trans_view_hook.unlink();
+  }
+  on_invalidated(t);
+}
+
+LogicalCachedExtent::~LogicalCachedExtent() {
+  if (has_parent_tracker() && is_valid() && !is_pending()) {
+    assert(get_parent_node());
+    auto parent = get_parent_node<FixedKVNode<laddr_t>>();
+    auto off = parent->lower_bound_offset(laddr);
+    assert(parent->get_key_from_idx(off) == laddr);
+    assert(parent->children[off] == this);
+    parent->children[off] = nullptr;
+  }
+}
+
+void LogicalCachedExtent::on_replace_prior(Transaction &t) {
+  assert(is_mutation_pending());
+  take_prior_parent_tracker();
+  assert(get_parent_node());
+  auto parent = get_parent_node<FixedKVNode<laddr_t>>();
+  //TODO: can this search be avoided?
+  auto off = parent->lower_bound_offset(laddr);
+  assert(parent->get_key_from_idx(off) == laddr);
+  parent->children[off] = this;
```

Comment:
cc @xxhdx1985126 

> Note that paddr<Dev(128),5369872384> is ONODE_BLOCK_STAGED and this was mutated before the get_pin().
But, the pin, the result of get_pin(), has a wrong parent pointer (seems that the parent pointer is not updated after a transaction is done).

See https://github.com/ceph/ceph/blob/1caadd7f2a642c1b490f3d8f4f1fb35f4684fcdc/src/crimson/os/seastore/btree/fixed_kv_node.h#L57-L70

There is an invariant that parent and children must **always** point to each other symmetricly before and after the commit. Seems your observation reflects this invarient is somehow broken. This should be the related logic to update the pointers between the LBALeaf and the LogicalCachedExtent(e.g. Onode) upon commit. Probably worth digging why the parent and children pointers are broken under the specific scenario.

> As a result, get_logical_child() accesses a child, which was invalid by the recent mutation (the prior pointer). I found that there is a condition to check the pointer is invalid, but just accessing it causes a seg fault.

Can you show the entire call stack about where the seg faulted check happens?

Reply:
I'm looking for a way to share the log. 

osd0.log: https://drive.google.com/file/d/1E3_wKRPKRLjIPVqSOT7b8due2yJiAYZV/view?usp=sharing
osd0.stdout: https://drive.google.com/file/d/1hpAakO-kimTbCQVxny933QJFbY4YsGcl/view?usp=sharing

Can you see the above logs (please understand the manual logs for debugging I added)?

Reply:
> Can you see the above logs (please understand the manual logs for debugging I added)?

Yes, I'm reading it.

Reply:
> osd0.log: https://drive.google.com/file/d/1E3_wKRPKRLjIPVqSOT7b8due2yJiAYZV/view?usp=sharing
> osd0.stdout: https://drive.google.com/file/d/1hpAakO-kimTbCQVxny933QJFbY4YsGcl/view?usp=sharing

@myoungwon What's the branch of the code? Is it main?

Reply:
Yes, it is main. I just added a few lines for logging (not related to seastore's operations. You can reproduce this by changing seastore_main_device_type in src/common/options/crimson.yaml.in to RANDOM_BLOCK_SSD. Then, run vstart.sh.

Reply:
@cyx1231st @myoungwon This issue should be fixed by #51948, please take a look